downloads/swan-lake-release-notes/swan-lake-2201.5.0/RELEASE_NOTE.md

```diff
@@ -0,0 +1,494 @@
+  for parameter or/and return types in generated Ballerina binding functions.
+
+    Consider the below Java methods and the corresponding Ballerina binding functions generated with and without the new command options.
+    ```java
+    // Parameters and return types having inbuilt object types (`java.lang.String`).
+    public String f1(String str) {
+        return str;
+    }
+    
+    // Parameter and return types having external object arrays (`Foo[]`).
+    public Foo[] f2(Foo[] fooArray) {
+        return fooArray;
+    }
+    ```
+    
+    **Without optional types** 
+    
+    ```bash
+    $ bal bindgen
+    ```
+    
+    ```ballerina
+    public function f1(string arg0) returns string {
+        return java:toString(Foo_strParamReturns(self.jObj, java:fromString(arg0))) ?: "";
+    }
+
+    public function f2(Foo[] arg0) returns Foo[]|error {
+        handle externalObj = Foo_objArrayParamReturns(self.jObj, check jarrays:toHandle(arg0, "Foo"));
+        Foo[] newObj = [];
+        handle[] anyObj = <handle[]>check jarrays:fromHandle(externalObj, "handle");
+        int count = anyObj.length();
+        foreach int i in 0 ... count - 1 {
+            Foo element = new (anyObj[i]);
+            newObj[i] = element;
+        }
+        return newObj;
+    }
+    ```
+    
+    **With optional types**
+
+    ```bash
+    $ bal bindgen --with-optional-types
+    ``` 
+    
+    ```ballerina
+    public function f1(string? arg0) returns string? {
+        return java:toString(Foo_strParamReturns(self.jObj, arg0 is () ? java:createNull() : java:fromString(arg0)));
+    }
+
+    public function f2(Foo?[]? arg0) returns Foo?[]?|error {
+        handle externalObj = Foo_objArrayParamReturns(self.jObj, check jarrays:toHandle(arg0 ?: [], "Foo"));
+        Foo?[]? newObj = [];
+        handle[] anyObj = <handle[]>check jarrays:fromHandle(externalObj, "handle");
+        int count = anyObj.length();
+        foreach int i in 0 ... count - 1 {
+            Foo? element = new (anyObj[i]);
+            if (newObj is Foo?[]) {
+                newObj[i] = element;
+            }
+        }
+        return newObj;
+    }
+    ```
+
+#### Language Server
+
+- Updated the LSP4J version to 0.15.0.
+- Added annotation completions for the tuple member context.
+- Added snippet completions for the `string` and `xml` templates.
+- Added code action support for regular expressions.
+- Added field access completions for template expressions.
+
+#### OpenAPI Tool
+- Updated the tool to handle multiple media types with the same return code in the Ballerina service to OpenAPI contract generation. 
+- Added the support to handle the newly introduced `@http:Query` annotation of the `ballerina/http` module in the Ballerina service to OpenAPI contract generation.
+- Added the support to generate API documentation for the resource functions in the generated Ballerina service. 
```

Comment:
```suggestion
- Added support to generate API documentation for the resource methods in the generated Ballerina service. 
```

Reply:
Is this an OpenAPI tool improvement?

Reply:
@NipunaRanasinghe

Reply:
@MaryamZi yes this is an improvement from the openAPI tool side to generate API documentation for resource methods, when generating services from its openAPI contract.

Reply:
@praneesha shall we modify this as `Added support to generate API documentation for the resource methods when generating Ballerina services using openAPI contracts`